Add Ingestion Job management API for Feast Core

mrzzy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/gojek/feast/blob/master/docs/contributing.md#code-conventions: 
2. Run unit tests and ensure that they are passing: https://github.com/gojek/feast/blob/master/docs/contributing.md#running-unit-tests
3. If your change introduces any API changes, make sure to update the integration tests scripts here: https://github.com/gojek/feast/tree/master/tests/e2e
4. Make sure documentation is updated for your PR!
5. Make sure you have signed the CLA https://cla.developers.google.com/clas

-->

**What this PR does / why we need it**:
PR to implement a Ingestion Job Management API into feast core.
This PR adds:
- extends CoreService API to  give the ability to list, stop, and restart ingestion jobs.
- extends Python sdk's `Client` to add support for the additions to the core service API.
- extends the CLI with `feast ingest-job list, describe, stop, restart`
- tests for the additions where appropriate.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

Fixes:  #302 

**Does this PR introduce a user-facing change?**: Yes
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information about release notes, see kubernetes' guide here:
http://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added Ingestion Job Management to Feast Core and the Python SDK:
- Listing Ingestion Jobs with filters
- Restarting Ingestion Jobs
- Stopping Ingestion Jobs

Added ingest-job subcommand to CLI for Ingestion Job Management:
-  feast ingest-job list, describe, stop, restart
```

woop commented on core/src/main/java/feast/core/service/JobService.java:
```diff
@@ -0,0 +1,223 @@
+   * @throws UnsupportedOperationException fsReference given is unsupported.
+   * @return Returns a list of matching featuresets
+   */
+  private List<FeatureSet> findFeatureSets(FeatureSetReference fsReference)
```
This should almost certainly be externalized. Can you have a look in the code base to see where else it occurs and see if we can reuse or generalize this?

mrzzy replied:
I moved the this method to `SpecService` as `matchFeatureSets()`.

zhilingc commented on core/src/main/java/feast/core/service/JobService.java:
```diff
@@ -0,0 +1,223 @@
+    // check job status is valid for stopping
+    Job job = getJob.get();
+    JobStatus status = job.getStatus();
+    if (status.equals(JobStatus.ABORTED)
```
You can use `JobStatus.TERMINAL_STATE` for this check

mrzzy replied:
How should calling `stopJob()` in one of the transitional states (ie`ABORTING`, `SUSPENDING` be handled? `JobStatus.TERMINAL_STATE` only the final states.

mrzzy replied:
I updated JobService use `JobStatus.TERMINAL_STATE` to check for final states. I added `JobStatus.TRANSITIONAL_STATES` to track transitional states (ie `PENDING`, `ABORTING`, `SUSPENDING`)

zhilingc commented on core/src/main/java/feast/core/dao/FeatureSetRepository.java:
```diff
@@ -36,6 +36,12 @@ FeatureSet findFirstFeatureSetByNameLikeAndProject_NameOrderByVersionDesc(
   List<FeatureSet> findAllByOrderByNameAscVersionAsc();
 
+  // find all feature sets by name and project name
+  List<FeatureSet> findAllByNameAndProject_Name(String name, String projectName);
```
I dont think you need these methods, might be better to inject the `SpecService` and use `ListFeatureSets`

mrzzy replied:
I refactored `JobService` to use `SpecService` to get its featuresets. The new methods have been removed.

zhilingc commented on protos/feast/core/IngestionJob.proto:
```diff
@@ -0,0 +1,65 @@
+}
+
+// Status of a Feast Ingestion Job
+enum IngestionJobStatus {
```
Could you possibly make feast.core.model.JobStatus a facade of this? or update the Job model to use this enum instead of our own

mrzzy replied:
I added `toIngestionProto()` to `JobStatus` that creates the `IngestionJobStatus` proto.

woop commented:
/ok-to-test

zhilingc commented on core/src/main/java/feast/core/service/JobService.java:
```diff
@@ -0,0 +1,228 @@
+   * @throws InvalidProtocolBufferException on error when constructing response protobuf
+   * @return list ingestion jobs response
+   */
+  public ListIngestionJobsResponse listJobs(ListIngestionJobsRequest request)
```
Since this method requires multiple separate calls to the db, it should be annotated with 
`@Transactional(readOnly=True)` to ensure that you're reading from the same snapshot of data

zhilingc commented on core/src/main/java/feast/core/service/SpecService.java:
```diff
@@ -131,6 +132,38 @@ public GetFeatureSetResponse getFeatureSet(GetFeatureSetRequest request)
   }
 
+  /**
+   * Finds &amp; returns the featuresets matching the given feature set reference. TODO: merge with
```
Yeah, you should merge this method. Add a private method to the jobservice to convert the feature reference to a filter if necessary.

mrzzy replied:
@woop wanted the method out of `JobService`, so I moved it to `SpecService`:
> This should almost certainly be externalized. Can you have a look in the code base to see where else it occurs and see if we can reuse or generalize this?
Where should the method be?

zhilingc replied:
You can use the existing method in specservice, just add a helper method to convert the `FeatureReference` to `Filter` since the former is a subset of the latter anyway.

mrzzy commented on core/src/main/java/feast/core/model/Job.java:
```diff
@@ -102,4 +106,31 @@ public void updateMetrics(List<Metrics> newMetrics) {
+   *
+   * @return Ingestion Job proto derieved from the given job
+   */
+  public IngestionJobProto.IngestionJob toIngestionProto() throws InvalidProtocolBufferException {
```
Fixed.

mrzzy commented on core/src/main/java/feast/core/model/JobStatus.java:
```diff
@@ -64,4 +66,38 @@
+   *
+   * @return IngestionJobStatus proto derieved from this job status
+   */
+  public IngestionJobStatus toIngestionProto() {
```
Fixed.

mrzzy commented on core/src/main/java/feast/core/model/JobStatus.java:
```diff
@@ -64,4 +66,38 @@
+   */
+  public IngestionJobStatus toIngestionProto() {
+    // maps job models job status to ingestion job status
+    Map<JobStatus, IngestionJobStatus> statusMap =
```
I think this should remain non static as it allows us to write: `status.toProto();`
instead of a more verbose 
`JobStatus.toProto(status);`

> assuming the signature of the method becomes `toProto(JobStatus)`

zhilingc replied:
No, I mean the statusMap should be static, not the method.

mrzzy commented:
/test test-end-to-end-batch

mrzzy commented:
/retest

mrzzy commented:
/test all

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/retest

mrzzy commented:
/test all

mrzzy commented:
/test all

mrzzy commented:
/test test-core-and-ingestion

mrzzy commented:
/test end-to-end

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-core-and-ingestion

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end-batch

mrzzy commented:
/test test-python-sdk

mrzzy commented:
/test test-core-and-ingestion

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end-batch

mrzzy commented:
/test test-python-sdk

mrzzy commented:
/retest

mrzzy commented:
This PR is non longer a work-in-progress and is ready for review.

I have manually tested to this to work with both Direct and Dataflow runners, although I like to ask a reviewer to double check that the integration with Dataflow is fully working.

woop commented:
> This PR is non longer a work-in-progress and is ready for review.

Thanks for this @mrzzy. Much needed

> I have manually tested to this to work with both Direct and Dataflow runners, although I like to ask a reviewer to double check that the integration with Dataflow is fully working.

How would the reviewer double check integration with Dataflow is working? Normally a reviewer only reviews the code, and the code should ideally contain all the necessary automation to test the functionality that is implemented.

mrzzy commented:
I agree that this is the ideal case, however the current end to end tests are only configured to test with the Direct runner, leaving a correctness blind spot with the Dataflow runner that has be addressed with manual testing.

woop commented:
> I agree that this is the ideal case, however the current end to end tests are only configured to test with the Direct runner, leaving a correctness blind spot with the Dataflow runner that has be addressed with manual testing.

Ideally the tests would be updated to incorporate Dataflow testing as well. We can possibly leave it out of scope for now. The reviewer in this case won't have time to actually spin up Dataflow and test this all out.

@mrzzy would you mind creating a new GitHub issue to add a Dataflow based test?

mrzzy commented:
I opened issue #595 for Dataflow Test Coverage.

zhilingc commented on core/src/main/java/feast/core/service/JobService.java:
```diff
@@ -0,0 +1,282 @@
+    // check job status is valid for restarting
+    Job job = getJob.get();
+    JobStatus status = job.getStatus();
+    if (JobStatus.getTransitionalStates().contains(status) || status.equals(JobStatus.UNKNOWN)) {
```
It might make sense to disallow restarting of jobs in terminal state as well, in order to prevent race conditions with the job coordinator. Another (smaller) issue with allowing job restarts for terminal jobs is that Dataflow only retains information about stopped jobs for a short period of time, after which the job cannot be retrieved any more.

mrzzy replied:
Thanks for pointing the race condition out, I have disabled restarting for terminal job statuses. As for your second point, this would not be a problem as the Dataflow Job Manager is configured to recreate the job once stopped.

zhilingc commented on sdk/python/feast/job.py:
```diff
@@ -187,3 +194,98 @@ def to_chunked_dataframe(
         return iter(self.result())
+
+
+class IngestJob:
```
Can you implement the __str__ method so that the object can be printed out in a human readable format? See feature_set.py:L87

mrzzy replied:
I have added `__str__()` and `__repr__()` to IngestJob to render it in a human readable format.

mrzzy commented:
/test test-core-and-ingestion

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end-batch

mrzzy commented:
/test test-python-sdk

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/retest

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/test test-end-to-end

mrzzy commented:
/retest

mrzzy commented:
/test test-core-and-ingestion

mrzzy commented:
/test test-end-to-end-batch

mrzzy commented:
/test test-python-sdk

mrzzy commented:
@zhilingc Thanks for the review and taking the time out of your day to test the dataflow runner.
I have implemented your comments into the PR:
-  disabled `stopJobs()` for suspended or aborted jobs due to possible race condition.
- added `__str__()` and `__repr__()` to `IngestJob` to render it in a human readable format.
- Added `feast ingest-job list, describe, stop, restart` to feast CLI
- Renamed `Job` to `RetrivalJob` to prevent confusion with `IngestJob` in python sdk.

zhilingc commented:
/lgtm
